src/kernel/emulation/linux/misc/sysctl_machdep.c

```diff
@@ -51,3 +80,139 @@ sysctl_handler(handle_brand_string)
 	return 0;
 }
 
+sysctl_handler(handle_max_basic)
+{
+	setup(0);
+
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+
+	sprintf(old,"%d",eax);
+
+	return 0;
+}
+
+sysctl_handler(handle_family)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax = eax >> 7;
+    eax &= 15;
+
+    sprintf(old,"%d",eax);
+
+    return 0;
+}
+
+sysctl_handler(handle_model)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax = eax >> 3;
+    eax &= 15;
+
+    __simple_sprintf(old,"%d",eax);
+
+    return 0;
+}
+
+sysctl_handler(handle_stepping)
+{
+    setup(1);
+
+    __cpuid(level, eax, ebx, ecx, edx);
+
+    eax &= 15;
+
+    sprintf(old,"%d",eax);
+
+    return 0;
+}
+
+sysctl_handler(handle_brand_string)
+{
+    setup(0x80000000);
+
+    __cpuid(level,eax,eabx, eacx, eadx);
+
+    if(eax < 0x80000004) // the information is not implemented
+        return 2;
+
+
+
+    union
+    {
+        unsigned int brand[12];
+        char name[49];
+    } v;
+
+		for (int i = 1; i < 4; i++)
+		{
+				eax = 0x80000000+i;
+				__cpuid(level, eax, ebx, ecx, edx);
+				v.brand[0x0+(i-1)*4] = eax;
+				v.brand[0x1+(i-1)*4] = ebx;
+				v.brand[0x2+(i-1)*4] = ecx;
+				v.brand[0x3+(i-1)*4] = edx;
+		}
+
+    v.name[48] = 0;
+    copyout_string(v.name, (char*) old, oldlen);
+
+    return 0;
+}
+sysctl_handler(handle_features)
+{
+    setup(1);
+
+    char *features[] = {"FPU","VME", "DE", "PSE", "TSC", "MSR", "PAE", "MCE", "CX8", "APIC", NULL,"SEP","MTRR","PGE",
+                        "MCA", "CMOV", "PSE-36", "PSN", "CLFSH", NULL, "DS", "ACPI", "MMX", "FXSR", "SSE", "SSE2", "SS",
+                        "HTT", "TM","IA64","PBE"};
+
+    char *out = ""; //"worst" case: we have all of it plus null termination
+
+    __cpuid(0,eax,ebx,ecx,edx);
+
+    int j = 0;
+    int counter = 0;
+
+    for (int i = 0; i < 32; i++)
+    {
+        if(edx>>i&1)
+        {
+						int len = __simple_strlen(features[i]);
+						int out_len = ____simple_strlen(out);
+						char *new_out = alloca((len+out_len+1)*sizeof(char))
```

Comment:
Don't `alloca` in a loop, that's a bad idea 'cause you're only going to free it when you exit the function.

`sizeof(char)` is always 1, it's not idiomatic to write it explicitly.

Why don't you just write the strings directly to the caller's buffer, instead of building the concatenated string on the stack and using `copyout` at the end?

Reply:
Yeah, good point, actually I was thinking that "how do I know that the `old` argument has enough memory allocated", but that should have that much if it is expected. Should I call `copyout_string` on it with an empty string first or it is a ~~very~~ bad idea?

Reply:
No, I wouldn't call `copyout_string` at all here. Just output these string pieces to `out` while there's enough space (if `out` is not null in the first place), and track the length yourself as you do now.

Reply:
Do you mean `old`? That is the actual output variable.

Reply:
Yes, I mean `old` of course.

Reply:
Okay, I'll do that. Thank you!